presto-main/src/main/java/com/facebook/presto/sql/planner/CanonicalPlanGenerator.java

```diff
@@ -88,6 +100,141 @@ public Optional<PlanNode> visitPlan(PlanNode node, Context context)
         return Optional.empty();
     }
 
+    @Override
+    public Optional<PlanNode> visitJoin(JoinNode node, Context context)
+    {
+        if (strategy == DEFAULT) {
+            return Optional.empty();
+        }
+        if (node.getType().equals(JoinNode.Type.RIGHT)) {
+            return visitJoin(node.flipChildren(), context);
+        }
+        List<PlanNode> sources = new ArrayList<>();
+        ImmutableList.Builder<RowExpression> allFilters = ImmutableList.builder();
+        ImmutableList.Builder<JoinNode.EquiJoinClause> criterias = ImmutableList.builder();
+        Stack<JoinNode> stack = new Stack<>();
```

Comment:
Why do we have a special treating of using stack here for consecutive joins instead of using the visitor for general cases?

Reply:
Other cases canoncialize plan nodes independently, more or less. Here, I'm gathering all join nodes and putting them in a single CanonicalJoinNode

Reply:
We have a similar use case for the ReorderJoins optimizer.  We use recursive function calls, but similar idea https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java#L520-L569.  I think the stack is fine, but maybe we should have the logic for creating the CanonicalJoinNode combined elsewhere so that it can be used for both CanonicalPlanGenerator and ReorderJoins. Unless the logic is different between the two.

Reply:
It isn't worth it. ReorderJoins has many different things like it limits how many nodes we want to join, has a determinismEvaluator(we dont really need it). It would be too congested to combine these implementation, given here it is a very simple implementation.